File: src/main/resources/templates/java/Lexer.java
package __JAVA_PACKAGE__;

/**
 * Token Manager.
 */
//@if(BASE_LEXER)
class Lexer extends __BASE_LEXER__ {
//@else
class Lexer {
//@fi

//@foreach(LOHI_BYTES)
    static final long[] jjbitVec__LOHI_BYTES_INDEX__ = { __LOHI_BYTES_VALUE__ };
//@end
//@foreach(STATES)
__STATES_NFA_AND_DFA__
//@end
    /**
     * Token literal values.
     */
    public static final String[] jjstrLiteralImages = {
    __LITERAL_IMAGES__};

    protected Token jjFillToken() {
        final Token t;
        final String curTokenImage;
//@if(KEEP_LINE_COOL)
        final int beginLine;
        final int endLine;
        final int beginColumn;
        final int endColumn;
//@fi
//@if(HAS_EMPTY_MATCH)
        if (jjmatchedPos < 0) {
            if (image == null) {
                curTokenImage = "";
            } else {
                curTokenImage = image.toString();
            }
//@if(KEEP_LINE_COOL)
            beginLine = endLine = input_stream.getEndLine();
            beginColumn = endColumn = input_stream.getEndColumn();
//@fi
        } else {
            String im = jjstrLiteralImages[jjmatchedKind];
            curTokenImage = (im == null) ? input_stream.GetImage() : im;
//@if(KEEP_LINE_COOL)
            beginLine = input_stream.getBeginLine();
            beginColumn = input_stream.getBeginColumn();
            endLine = input_stream.getEndLine();
            endColumn = input_stream.getEndColumn();
//@fi
        }
//@else
        String im = jjstrLiteralImages[jjmatchedKind];
        curTokenImage = (im == null) ? input_stream.GetImage() : im;
//@if(KEEP_LINE_COOL)
        beginLine = input_stream.getBeginLine();
        beginColumn = input_stream.getBeginColumn();
        endLine = input_stream.getEndLine();
        endColumn = input_stream.getEndColumn();
//@fi
//@fi
        t = new Token(jjmatchedKind, curTokenImage);
//@if(KEEP_LINE_COOL)
        t.beginLine = beginLine;
        t.endLine = endLine;
        t.beginColumn = beginColumn;
        t.endColumn = endColumn;
//@fi
        return t;
    }

//@invoke(DUMP_STATE_SETS)
//@foreach(NON_ASCII_TABLE)
    private static final boolean jjCanMove___NON_ASCII_TABLE_METHOD__(int hiByte, int i1, int i2, long l1, long l2) {
        switch (hiByte) {
            __NON_ASCII_TABLE_MOVE__
            return false;
        }
    }

//@end

    int curLexState = __DEFAULT_LEX_STATE__;
    int defaultLexState = __DEFAULT_LEX_STATE__;
    int jjnewStateCnt;
    int jjround;
    int jjmatchedPos;
    int jjmatchedKind;

/** Get the next Token. */
    public Token getNextToken() {
//@if(HAS_SPECIAL)
        Token specialToken = null;
//@fi
        Token matchedToken;
        int curPos = 0;

EOFLoop :
        for (;;) {
            try {
                curChar = input_stream.BeginToken();
            } catch(Exception e) {
//@if(DEBUG_TOKEN_MANAGER)
                debugStream.println(\"Returning the <EOF> token.\\n\");
//@fi
                jjmatchedKind = 0;
                jjmatchedPos = -1;
                matchedToken = jjFillToken();
//@if(HAS_SPECIAL)
                matchedToken.specialToken = specialToken;
//@fi
//@invoke(DUMP_GET_NEXT_TOKEN)
        }
    }

//@if(DEBUG_TOKEN_MANAGER)
    protected static final int[][][] statesForState = __STATES_FOR_STATE__;
    protected static final int[][] kindForState = __KIND_FOR_STATE__;
//@fi
//@if(HAS_LOOP)
	int[] jjemptyLineNo = new int[__MAX_LEX_STATES__];
	int[] jjemptyColNo = new int[__MAX_LEX_STATES__];
    boolean[] jjbeenHere = new boolean[__MAX_LEX_STATES__];
//@fi
    void SkipLexicalActions(Token matchedToken) {
        switch (jjmatchedKind) {
//@invoke(DUMP_SKIP_ACTIONS)
            default:
                break;
        }
    }

    void MoreLexicalActions() {
        jjimageLen += (lengthOfMatch = jjmatchedPos + 1);
        switch (jjmatchedKind) {
//@invoke(DUMP_MORE_ACTIONS)
            default:
                break;
        }
    }

    void TokenLexicalActions(Token matchedToken) {
        switch (jjmatchedKind) {
//@invoke(DUMP_TOKEN_ACTIONS)
            default:
                break;
        }
    }

    private void jjCheckNAdd(int state) {
        if (jjrounds[state] != jjround) {
            jjstateSet[jjnewStateCnt++] = state;
            jjrounds[state] = jjround;
        }
    }

    private void jjAddStates(int start, int end) {
        do {
            jjstateSet[jjnewStateCnt++] = jjnextStates[start];
        } while (start++ != end);
    }

    private void jjCheckNAddTwoStates(int state1, int state2) {
        jjCheckNAdd(state1);
        jjCheckNAdd(state2);
    }

//@if(CHECK_NADD_STATES_DUAL_NEEDED)
    private void jjCheckNAddStates(int start, int end) {
        do {
            jjCheckNAdd(jjnextStates[start]);
        } while (start++ != end);
    }

//@fi
//@if(CHECK_NADD_STATES_UNARY_NEEDED)
    private void jjCheckNAddStates(int start) {
        jjCheckNAdd(jjnextStates[start]);
        jjCheckNAdd(jjnextStates[start + 1]);
    }

//@fi

    /**
     * Constructor.
     */
    public Lexer(JavaCharStream stream) {
        input_stream = stream;
    }

    /**
     * Constructor.
     */
    public Lexer(JavaCharStream stream, int lexState) {
        ReInit(stream);
        SwitchTo(lexState);
    }

    /**
     * Reinitialise parser.
     */
    public void ReInit(JavaCharStream stream) {
        jjmatchedPos = 0;
        jjnewStateCnt = 0;
        curLexState = defaultLexState;
        input_stream = stream;
        ReInitRounds();
    }

    private void ReInitRounds() {
        int i;
        jjround = 0x80000001;
        for (i = __STATE_SET_SIZE__; i-- > 0; ) {
            jjrounds[i] = 0x80000000;
        }
    }

    /**
     * Reinitialise parser.
     */
    public void ReInit(JavaCharStream stream, int lexState) {
        ReInit(stream);
        SwitchTo(lexState);
    }

    /**
     * Switch to specified lex state.
     */
    public void SwitchTo(int lexState) {
        if (lexState >= __STATE_COUNT__ || lexState < 0)
            throw new TokenException(
                "Error: Ignoring invalid lexical state : " + lexState + ". State unchanged.",
                TokenException.INVALID_LEXICAL_STATE);
        else
            curLexState = lexState;
    }

    /**
     * Lexer state names.
     */
    public static final String[] lexStateNames = {
//@foreach(STATE_NAMES)
        "__STATE_NAMES_VALUE__",
//@end
    };
//@invoke(DUMP_STATIC_VAR_DECLARATIONS)

    private       JavaCharStream input_stream;
    private final int[]          jjrounds   = new int[__STATE_SET_SIZE__];
    private final int[]          jjstateSet = new int[2 * __STATE_SET_SIZE__];
    private final StringBuilder  jjimage    = new StringBuilder();
    private       StringBuilder  image      = jjimage;
    private       int            jjimageLen;
    private       int            lengthOfMatch;
    protected     int            curChar;
}

File: src/main/resources/templates/rust/lexer.rs
use std::cmp;
use crate::__RUST_MODULE__::token::Token;
use crate::__RUST_MODULE__::charstream::CharStream;

//@foreach(LOHI_BYTES)
const JJBIT_VEC__LOHI_BYTES_INDEX__ : [u64; __LOHI_BYTES_LENGTH__] = [__LOHI_BYTES_BYTES__];
//@end

pub const LEX_STATE_NAMES : [&str; __STATE_NAMES_LENGTH__] = [
//@foreach(STATE_NAMES)
   "__STATE_NAME__",
//@end
];

pub const JJSTR_LITERAL_IMAGES : [&str; __LITERAL_IMAGES_LENGTH__] = [
//@foreach(LITERAL_IMAGES)
   "__LITERAL_IMAGE_NAME__",
//@end
];

//@var(dumpStaticVarDeclarations)

//@var(DumpStateSets)


//@if(DEBUG_TOKEN_MANAGER)
const statesForState : [][][] = {__STATES_FOR_STATE__};
const kindForState : [][] = {__KIND_FOR_STATE__};
//@fi

pub struct Lexer<'a> {
  cur_lex_state: i8,
  default_lex_state: i8,
  jjnew_state_cnt: usize,
  jjround: usize,
  jjmatched_pos: usize,
  jjmatched_kind: u32,

  jjrounds : [usize; __STATE_SET_SIZE__],
  jjstate_set : [usize; __STATE_SET_SIZE2__],
  jjimage_len : usize,
  length_of_match : usize,
  cur_char : u32,
  input_stream: CharStream<'a>,
//@if(HAS_LOOP)
  jjemptyLineNo : [usize; __MAX_LEX_STATES__],
  jjemptyColNo : [usize; __MAX_LEX_STATES__],
  jjbeenHere : [bool; __MAX_LEX_STATES__],
//@fi

}

impl<'a> Lexer<'a> {

  pub fn new(text: &'a str) -> Self {
    Lexer {
      cur_lex_state: __DEFAULT_LEX_STATE__,
      default_lex_state: __DEFAULT_LEX_STATE__,
      jjnew_state_cnt: 0,
      jjround: 0,
      jjmatched_pos: 0,
      jjmatched_kind: 0,
      jjrounds: [0; __STATE_SET_SIZE__],
      jjstate_set: [0; __STATE_SET_SIZE2__],
      jjimage_len: 0,
      length_of_match: 0,
      cur_char: 0,
      input_stream: CharStream::new(text),
//@if(HAS_LOOP)
      jjemptyLineNo: [0, __MAX_LEX_STATES__],
      jjemptyColNo: [0, __MAX_LEX_STATES__],
      jjbeenHere: [false, __MAX_LEX_STATES__],
//@fi
    }
  }

  pub fn new_from_state(text: &'a str, lex_state: i8) -> Self {
    let mut lexer = Lexer {
      cur_lex_state: lex_state,
      default_lex_state: lex_state,
      jjnew_state_cnt: 0,
      jjround: 0,
      jjmatched_pos: 0,
      jjmatched_kind: 0,
      jjrounds: [0; __STATE_SET_SIZE__],
      jjstate_set: [0; __STATE_SET_SIZE2__],
      jjimage_len: 0,
      length_of_match: 0,
      cur_char: 0,
      input_stream: CharStream::new(text),
//@if(HAS_LOOP)
      jjemptyLineNo: [0, __MAX_LEX_STATES__],
      jjemptyColNo: [0, __MAX_LEX_STATES__],
      jjbeenHere: [false, __MAX_LEX_STATES__],
//@fi
    };
    lexer.re_init_rounds();
    lexer.switch_to(lex_state);
    lexer
  }
/*
  pub TokenManager(JavaCharStream stream) {
    input_stream = stream;
  }
*/

  pub fn switch_to(&mut self, lex_state: i8) {
    if lex_state >= __STATE_COUNT__ {
       panic!("Error: Ignoring invalid lexical state : {}. State unchanged." , lex_state);
    }
    else {
      self.cur_lex_state = lex_state;
    }
  }

  fn re_init_rounds(&mut self)
  {
    let mut i : usize = __STATE_SET_SIZE__;
    self.jjround = 0x80000001;
    while i > 0 {
        i -= 1;
      self.jjrounds[i] = 0x80000000;
    }
  }

  fn jj_fill_token(&self) -> Token {
//@if(KEEP_LINE_COOL)
    let mut begin_line : usize = 0;
    let mut end_line : usize = 0;
    let mut begin_column : usize = 0;
    let mut end_column : usize = 0;
//@fi
//@if(HAS_EMPTY_MATCH)
    if self.jjmatched_pos < 0 {
      if (image == null)
        cur_token_image = "";
      else
        cur_token_image = image.toString();

//@if(KEEP_LINE_COOL)
      begin_line = end_line = self.input_stream.get_end_line();
      begin_column = end_column = self.input_stream.get_end_column();
//@fi
    } else {
      String im = JJSTR_LITERAL_IMAGES[self.jjmatched_kind as usize];
      cur_token_image: &str = if im == "" { self.input_stream.get_image() } else { im };
//@if(KEEP_LINE_COOL)
      begin_line = self.input_stream.get_begin_line();
      begin_column = self.input_stream.get_begin_column();
      end_line = self.input_stream.get_end_line();
      end_column = self.input_stream.get_end_column();
//@fi
    }
//@else
    let im : &str = JJSTR_LITERAL_IMAGES[self.jjmatched_kind as usize];
    let cur_token_image= if im == "" { self.input_stream.get_image() } else { im.to_string() };
//@if(KEEP_LINE_COOL)
    begin_line = self.input_stream.get_begin_line();
    begin_column = self.input_stream.get_begin_column();
    end_line = self.input_stream.get_end_line();
    end_column = self.input_stream.get_end_column();
//@fi
//@fi
//@if(KEEP_LINE_COOL)
    Token::new(self.jjmatched_kind, cur_token_image, begin_line, begin_column, end_line, end_column)
//@else
    Token::new(self.jjmatched_kind, cur_token_image)
//@fi
  }

  pub fn get_next_token(&mut self) -> Token {
//@if(HAS_SPECIAL)
    let mut specialToken: Token;
//@fi
    let mut matched_token: Token;
    let mut cur_pos: usize = 0;

    'EOFLoop:
    loop {
      let result = self.input_stream.begin_token();
      if !result.is_err() {
        self.cur_char = u32::from(result.unwrap());
      } else {
//@if(DEBUG_TOKEN_MANAGER)
        debugStream.println(\"Returning the <EOF> token.\\n\");
//@fi
        self.jjmatched_kind = 0;
        self.jjmatched_pos = usize::MAX;
        matched_token = self.jj_fill_token();
//@if(HAS_SPECIAL)
        matched_token.specialToken = specialToken;
//@fi
        __DumpGetNextToken__
    }
    Token::new(0, String::from(""), 0,0,0, 0)
  }

  fn skip_lexical_actions(&self, matched_token: &Token) {
    match self.jjmatched_kind {
      __DumpSkipActions__
      _=> {}
    }
  }

  fn more_lexical_actions(&mut self) {
    self.length_of_match = self.jjmatched_pos + 1;
    self.jjimage_len += self.length_of_match;
    match self.jjmatched_kind {
      __DumpMoreActions__
      _=> {}
    }
  }

  fn token_lexical_actions(&self, matched_token: &Token) {
    match self.jjmatched_kind {
      __DumpTokenActions__
      _=> {}
    }
  }

  fn jj_check_n_add(&mut self, state: usize) {
    if self.jjrounds[state] != self.jjround {
      self.jjstate_set[self.jjnew_state_cnt] = state;
      self.jjnew_state_cnt += 1;
      self.jjrounds[state] = self.jjround;
    }
  }

  fn jj_add_states(&mut self, start: usize, end: usize) {
    let mut do_while = true;
    let mut index = start;
    while do_while {
      self.jjstate_set[self.jjnew_state_cnt] = JJNEXT_STATES[index];
      self.jjnew_state_cnt += 1;
      do_while = index != end;
      index += 1;
    }
  }

  fn jj_check_n_add_two_states(&mut self, state1: usize, state2: usize) {
    self.jj_check_n_add(state1);
    self.jj_check_n_add(state2);
  }
//@if(CHECK_NADD_STATES_DUAL_NEEDED)

  fn jj_check_n_add_states(&mut self, start: usize, end: usize) {
    let mut do_while = true;
    let mut index = start;
    while do_while {
      self.jj_check_n_add(JJNEXT_STATES[index]);
      do_while = index != end;
      index += 1;
    }
  }
//@fi
//@if(CHECK_NADD_STATES_UNARY_NEEDED)

  fn jj_check_n_add_states(&mut self, start: usize) {
    self.jj_check_n_add(JJNEXT_STATES[start]);
    self.jj_check_n_add(JJNEXT_STATES[start + 1]);
  }
//@fi

//@foreach(STATES)
__NfaAndDfa__
//@end
}

//@foreach(NON_ASCII_TABLE)
fn jj_can_move__ASCII_METHOD__(hi_byte: u32, i1: usize, i2: usize, l1: u64, l2: u64) -> bool {
  match hi_byte {
__ASCII_MOVE__
  }
}

//@end